requests-for-comments/2025-02-25-query-migrations.md

```diff
@@ -0,0 +1,80 @@
+# Request for comments: Query Migrations
+
+## Problem statement
+
+We'd like to evolve the [query schema](https://github.com/PostHog/posthog/blob/master/frontend/src/queries/schema/index.ts) while keeping backwards compatibility for existing queries.
+
+## Context
+
+Queries are persisted in various places:
+
+1. The `query` field of the Insights model.
+2. Local storage for the "query draft" feature.
+3. The url for shareable insight links.
+4. Notebook nodes (and canvas element).
+5. The activity log, when changing a query.
+6. Queries that are converted from `filters` e.g. when users create an insight by directly calling the api.
+7. ...? _(let me know)_
+
+Some of these places are available frontend-side (e.g. queries in a url) and some are available backend-side (e.g. queries in an insight model).
+
+This makes it difficult to migrate a query with a non-backwards compatibly change e.g. converting a boolean to an enum. Backward compatible changes e.g. adding an optional field are no problem.
+
+Any errors in a query cause Pydantic to fail the validation and thus insights with the error will fail to compute.
+
+## Design
+
+Migrating the schema from one version to the next should be as easy as a database migration for developers.
+
+### Option a) Add a backend side migration and allow the frontend to use this in a new endpoint
+
+We could add only a backend side migration and let the frontend call a `/query/validation` endpoint that also migrates queries to newer versions.
```

Comment:
Since queries are always retrieved from the backend, in most cases we can do something sneakier that avoids the roundtrip: In all of the call sites from the "Queries are persisted in various places" list, we'd do the `/query/validation` transformation internally before the query ever gets to the user. This would mean the frontend _never_ seeing old-style queries.

With one exception: filters in the URL  – where the roundtrip would still be necessary, and would be quite painful (e.g. in some cases the frontend would not be able to support the query anymore, so it would _have_ to wait for the validation round trip – a few hundred milliseconds of a wait doesn't sound so bad, but it can definitely add up to a slow experience!)

Reply:
Yup, we definitely want to handle this backend side as much as possible. I'll probably find out how easy that is for notebooks, activity logs, etc when working on it 😆 

We also have queries stored in local storage, so these need to do the roundtrip as well.

Reply:
Why would migrating the query take few hundred ms? Correct me if wrong, but the `/query/validation` should parse => apply migrations => serialize, it should not take more than 50-100ms?

It would also be easy to cache those.

Reply:
Validation itself should be quick, but there's also significant overhead to every individual API request, plus internet speeds vary – piece by piece these things add up to a slow app experience, translating to lower retention → less 💵.